File: android/app/src/main/java/com/liive/ride/designsystem/LiiveSosButton.kt
//  LiiveSosButton.kt  ·  Liive Ride DS (Compose)  ·  mirrors components/ride/SOSButton
//  Coral disc + continuous pulse halo + press-shrink. onActivate should show a
//  confirm dialog before contacting emergency services.
package com.liive.ride.designsystem

import androidx.compose.animation.core.*
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.interaction.collectIsPressedAsState
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.Text
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.scale
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.sp

@Composable
fun LiiveSosButton(
    size: Dp = LiiveControl.xl + LiiveSpacing.s,
    showLabel: Boolean = true,
    modifier: Modifier = Modifier,
    onActivate: () -> Unit
) {
    val c = LiiveTheme.colors
    val interaction = remember { MutableInteractionSource() }
    val pressed by interaction.collectIsPressedAsState()
    val pressScale by animateFloatAsState(
        targetValue = if (pressed) LiiveSosButtonLayout.PressedScale else LiiveSosButtonLayout.RestingScale,
        animationSpec = tween(durationMillis = LiiveMotion.fastMs, easing = LiiveMotion.easeOut),
        label = "LiiveSosButtonPressScale"
    )
    val transition = rememberInfiniteTransition(label = "sos")
    val pulse by transition.animateFloat(
        initialValue = LiiveSosButtonLayout.PulseStart,
        targetValue = LiiveSosButtonLayout.PulseEnd,
        animationSpec = infiniteRepeatable(
            tween(LiiveSosButtonLayout.PulseDurationMs, easing = LiiveMotion.easeOut),
            RepeatMode.Restart
        ),
        label = "pulse"
    )
    Box(
        modifier
            .size(size)
            .semantics { contentDescription = "Emergency SOS" },
        contentAlignment = Alignment.Center
    ) {
        Box(
            Modifier.matchParentSize().graphicsLayer {
                val pulseScale = LiiveSosButtonLayout.RestingScale + pulse * LiiveSosButtonLayout.PulseScaleDelta
                scaleX = pulseScale
                scaleY = pulseScale
                alpha = (LiiveSosButtonLayout.PulseEnd - pulse) * LiiveSosButtonLayout.PulseStartOpacity
            }.clip(CircleShape).background(c.danger)
        )
        Box(
            Modifier.matchParentSize().scale(pressScale)
                .shadow(
                    elevation = LiiveElevation.sos,
                    shape = CircleShape,
                    clip = false,
                    ambientColor = c.danger.copy(alpha = LiiveSosButtonLayout.ShadowColorAlpha),
                    spotColor = c.danger.copy(alpha = LiiveSosButtonLayout.ShadowColorAlpha),
                )
                .clip(CircleShape)
                .background(c.danger)
                .clickable(interactionSource = interaction, indication = null) { onActivate() },
            contentAlignment = Alignment.Center
        ) {
            Column(
                horizontalAlignment = Alignment.CenterHorizontally,
                verticalArrangement = Arrangement.spacedBy(LiiveSosButtonLayout.LabelGap)
            ) {
                Text(
                    "SOS",
                    color = LiiveSosButtonLayout.ForegroundColor,
                    fontWeight = FontWeight.Bold,
                    fontSize = (size.value * LiiveSosButtonLayout.SosTextScale).sp,
                    fontFamily = SchibstedGrotesk
                )
                if (showLabel) {
                    Text(
                        "HELP",
                        color = LiiveSosButtonLayout.ForegroundColor.copy(alpha = LiiveSosButtonLayout.HelpTextOpacity),
                        fontWeight = FontWeight.SemiBold,
                        fontSize = (size.value * LiiveSosButtonLayout.HelpTextScale).sp,
                        fontFamily = SchibstedGrotesk,
                        letterSpacing = LiiveSosButtonLayout.HelpLetterSpacing
                    )
                }
            }
        }
    }
}

private object LiiveSosButtonLayout {
    val LabelGap = LiiveSpacing.xs2 / 2
    const val RestingScale = 1f
    const val PressedScale = 0.94f
    const val PulseStart = 0f
    const val PulseEnd = 1f
    const val PulseScaleDelta = 0.5f
    const val PulseStartOpacity = 0.35f
    const val PulseDurationMs = LiiveMotion.slowMs + LiiveMotion.slowMs + LiiveMotion.slowMs +
        LiiveMotion.baseMs + LiiveMotion.fastMs
    const val SosTextScale = 0.28f
    const val HelpTextScale = 0.15f
    const val HelpTextOpacity = 0.9f
    const val ShadowColorAlpha = 0.45f
    val HelpLetterSpacing = (LiiveSpacing.xs2.value / 4).sp
    val ForegroundColor = Color.White
}

File: Packages/RideSharingFeature/Sources/RideSharingFeature/DesignSystem/LiiveSOSButton.swift
//  LiiveSOSButton.swift  ·  Liive Ride DS (SwiftUI)
//  Always-reachable emergency control: coral disc + continuous pulse halo,
//  press-shrink. onActivate should present a confirm dialog before dialing.

import SwiftUI

public struct LiiveSOSButton: View {
    var size: CGFloat = LiiveControl.xl + LiiveSpacing.s
    var showLabel: Bool = true
    let onActivate: () -> Void

    @State private var pressed = false
    @State private var pulse = false

    public init(
        size: CGFloat = LiiveControl.xl + LiiveSpacing.s,
        showLabel: Bool = true,
        onActivate: @escaping () -> Void
    ) {
        self.size = size; self.showLabel = showLabel; self.onActivate = onActivate
    }

    public var body: some View {
        Button(action: onActivate) {
            ZStack {
                Circle()
                    .fill(LiiveColor.danger)
                    .scaleEffect(pulse ? LiiveSOSButtonLayout.pulseScale : LiiveSOSButtonLayout.restingScale)
                    .opacity(pulse ? LiiveSOSButtonLayout.pulseEndOpacity : LiiveSOSButtonLayout.pulseStartOpacity)
                Circle().fill(LiiveColor.danger).liiveShadow(.sos)
                VStack(spacing: LiiveSOSButtonLayout.labelGap) {
                    Text("SOS").font(.system(size: size * LiiveSOSButtonLayout.sosTextScale, weight: .bold, design: .rounded))
                    if showLabel {
                        Text("HELP").font(.system(size: size * LiiveSOSButtonLayout.helpTextScale, weight: .semibold, design: .rounded))
                            .foregroundColor(.white.opacity(LiiveSOSButtonLayout.helpTextOpacity))
                            .tracking(LiiveSOSButtonLayout.helpLetterSpacing)
                    }
                }
                .foregroundColor(LiiveSOSButtonLayout.foregroundColor)
            }
            .frame(width: size, height: size)
            .scaleEffect(pressed ? LiiveSOSButtonLayout.pressedScale : LiiveSOSButtonLayout.restingScale)
            .animation(.easeOut(duration: LiiveMotion.fast), value: pressed)
        }
        .buttonStyle(.plain)
        .accessibilityLabel(Text("Emergency SOS"))
        .simultaneousGesture(DragGesture(minimumDistance: 0)
            .onChanged { _ in pressed = true }.onEnded { _ in pressed = false })
        .onAppear {
            withAnimation(.easeOut(duration: LiiveSOSButtonLayout.pulseDuration).repeatForever(autoreverses: false)) { pulse = true }
        }
    }
}

private enum LiiveSOSButtonLayout {
    static let labelGap = LiiveSpacing.xs2 / 2
    static let restingScale: CGFloat = 1.0
    static let pressedScale: CGFloat = 0.94
    static let pulseScale: CGFloat = 1.5
    static let pulseStartOpacity = 0.35
    static let pulseEndOpacity = 0.0
    static let pulseDuration = LiiveMotion.slow + LiiveMotion.slow + LiiveMotion.slow + LiiveMotion.base + LiiveMotion.fast
    static let sosTextScale: CGFloat = 0.28
    static let helpTextScale: CGFloat = 0.15
    static let helpTextOpacity = 0.9
    static let helpLetterSpacing = LiiveSpacing.xs2 / 4
    static let foregroundColor = Color.white
}
